pkg/util/watch.go

```diff
@@ -374,20 +374,41 @@ func Watch[C WatchClient[L], L metav1.ListMetaAccessor, T any, P WatchObject[T]]
 				relistItems := accessors.Items(relistList)
 
 				func() {
-					// First, copy the contents of objects into oldObjects. We do this so that we can
-					// uphold some guarantees about the contents of the store.
+					store.mutex.Lock()
+					defer store.mutex.Unlock()
+
+					// Copy the current contents of objects, and start tracking which ones have
+					// since been deleted.
 					oldObjects := make(map[types.UID]*T)
+					deleted := make(map[types.UID]bool)
```

Comment:
I tend to avoid maps to bools because it's often unclear what `deleted[uid] = false` is supposed to mean (and intent is hard to discern as a reader.

Reply:
Makes sense. updated